💄 additional active icon states

iantrich opened the pull request:
I think highlighting devices that require potential attention is the main goal and after some thought propose the following changes:

**Domain - Active State**
_Existing_
```
light - on
switch - on
binary_sensor - on
fan - on
sun - above_horizon
```
_Additional_
```
input_boolean - on
cover - open
media_player - playing/paused
device_tracker/person - home
plant - problem
lock - unlocked
climate - heating/cooling
```

fixes https://github.com/home-assistant/home-assistant-polymer/issues/3897

iantrich committed:
💄 additional active icon states

iantrich committed:
♻️ sort by domain

ludeeus commented on src/components/entity/state-badge.ts:
```diff
@@ -111,12 +111,21 @@ class StateBadge extends LitElement {
+      ha-icon[data-domain="light"][data-state="on"],
+      ha-icon[data-domain="input_boolean"][data-state="on"],
+      ha-icon[data-domain="lock"][data-state="unlocked"],
+      ha-icon[data-domain="media_player"][data-state="paused"],
```
This is not an active state?

iantrich replied:
I disagree. Something is actively paused. The device is still on and attention should be drawn to it over idle/off

ludeeus commented on src/components/entity/state-badge.ts:
```diff
@@ -111,12 +111,21 @@ class StateBadge extends LitElement {
-      ha-icon[data-domain="sun"][data-state="above_horizon"] {
+      ha-icon[data-domain="light"][data-state="on"],
+      ha-icon[data-domain="input_boolean"][data-state="on"],
+      ha-icon[data-domain="lock"][data-state="unlocked"],
```
This is the opposite of the active state of a lock.
When locked the lock should have an active icon.

iantrich replied:
I disagree. Active happens to be the theme var name, and I didn't want to change that. As stated in my description, this is about drawing attention. A locked door is not something to have attention drawn to. Just like a closed door shouldn't.

ludeeus commented on src/panels/lovelace/cards/hui-entity-button-card.ts:
```diff
@@ -207,11 +207,21 @@ class HuiEntityButtonCard extends LitElement implements LovelaceCard {
-      ha-icon[data-domain="sun"][data-state="above_horizon"] {
+      ha-icon[data-domain="light"][data-state="on"],
+      ha-icon[data-domain="input_boolean"][data-state="on"],
+      ha-icon[data-domain="lock"][data-state="unlocked"],
```
This is the opposite of the active state of a lock.
When locked the lock should have an active icon.

ludeeus commented on src/panels/lovelace/cards/hui-entity-button-card.ts:
```diff
@@ -207,11 +207,21 @@ class HuiEntityButtonCard extends LitElement implements LovelaceCard {
+      ha-icon[data-domain="light"][data-state="on"],
+      ha-icon[data-domain="input_boolean"][data-state="on"],
+      ha-icon[data-domain="lock"][data-state="unlocked"],
+      ha-icon[data-domain="media_player"][data-state="paused"],
```
This is not an active state?

ludeeus reviewed (commented):
I feel like paused for media players should be removed, the media player does not ha an active state when it's not doing anything.

As for the lock, I get why you had it like this, but it should be inverted to conform with other active states.

balloob reviewed (changes requested):
My comment from linked PR - https://github.com/home-assistant/home-assistant-polymer/issues/3897#issuecomment-483344529:

> Each time this comes up, I say the same thing: don't suggest that we need to color this or that, instead, come up with a strategy that we can use for each place (entity row, entity button etc) on how we should color each type of entity/device class and each of their states.

PeteBa commented:
Looks like there is two competing strategies here. The first being **highlight when active** (e.g. "locked") the other being **highlight when requires attention** (e.g. "unlocked"). Feels like that choice needs to be made first.

Personal preference is the former as it can be explicitly coded. The difficulty with "requires attention" is that it varies with context e.g. a locked fire-exit door requires attention.

iantrich commented:
@balloob I believe I did, my strategy is highlighting items requiring potential attention. Or are you saying I should explicitly state for each and every entity/state combination? I think that's overkill and this can evolve overtime. 

I don't think there should be a difference between state badge and entity button.

balloob commented:
I think that if something requires attention, like a plant in "problem" state, we should color it red or even add a red badge to the icon with an `!` in it.

balloob commented:
So I'm glad that we're talking about this issue btw, and let's find a solution 👍 and sorry for not reading the description thoroughly when leaving my review. 

So I don't know if "potentially requires attention" is a good strategy. Is a person being home requiring attention ? Wouldn't either "active" or "open" be a better strategy?

arsaboo commented:
Will it be possible (with this PR) to allow little more user control? For example, we may not know what requires attention. Also, attention may not be a simple binary state, i.e., there are different levels of attention.

balloob commented:
What is the driver of this change? Right now glance and picture elements use the same component as the entities row card. Glance and picture elements could benefit of such changes, entity row not (it would become too busy).

@arsaboo that is out of scope of this pr

iantrich commented:
@balloob person was admittedly something I struggled to fit within this reasoning. Let me take sometime to better describe each case and we'll go from there. I really want to get something resolved on this as well 😉

iantrich commented:
Also,
I believe all those elements benefit from state-badge changes. I'll double check

balloob commented:
The problem is that if everything wants to grab attention, nothing will.

balloob commented:
The problem is that if everything wants to grab attention, nothing will.

iantrich commented:
So my idea is that we should only highlight/change an icon color when it warrants attention which can be a very tricky thing to balance as @balloob noted. My initial changes were the ones I saw most value from, but if I'm doing this, I might as well do it right and consider ALL domains, so I'll try my best to explain each domain below (sourced from domain_icon.ts, so may not be complete) and hopefully we can finish with something that makes a lot of people happy:

-  alarm_control_panel: **armed**
_When an alarm is armed, your home is typically restricted in some way. e.g. notifications on motion which I believe warrants more attention from a user. As you'll see below, this somewhat contends with other ideas on things like doors/locks, but the fact that an armed home is usually used to restrict a home, I think it warrants greater attention than disarmed._
-   alert: **on**
_I don't think there will be much discussion for this one._
-   alexa: **(I'm not sure what this represents, actually. The connection?)**
-   automation: **(I'm undecided if this one warrants any change or not)**
-   binary_sensor: **on**
_I don't think there will be much discussion for this one._
-   calendar: **on**
_I believe this is only on during an event, and if so would make sense to be when it is highlighted_
-   camera: **streaming**
_Not really sure about this one yet, I'm conflicted_
-   climate: **heating/cooling**
_Your furnace actively pushing air and thereby using more energy warrants more attention_
-   configurator: **(Should not change)**
-   conversation: **(Should not change)**
-   counter: **(Should not change)**
-   cover: open
_I don't think there will be much discussion for this one._
-   device_tracker/person: **home**
_I've went back and forth on this one a bit, but I believe that home is of more significance than away. Change my mind :smile:_
-   fan: **on**
_I don't think there will be much discussion for this one._
-   google_assistant: **(I'm not sure what this represents, actually. The connection?)**
-   group: **(This can take on numerous states, dependent on what type of entities it consists of. I don't have a clear answer for this.)**
-   history_graph: **(A relic of states UI, correct?)**
-   homeassistant: **(I'm not sure what this represents)**
-   homekit: **(I'm not sure what this represents, actually. The connection?)**
-   image_processing: **(I'm not familiar enough with this integration)**
-   input_boolean: **on**
_I don't think there will be much discussion for this one._
-   input_datetime: **(Should not change)**
-   input_number: **(Should not change)**
-   input_select: **(Should not change)**
-   input_text: **(Should not change)**
-   light: **on**
_I don't think there will be much discussion for this one._
-   lock: unlocked
_Along the same lines as a garage door or door sensor, an unlocked door makes for a less secure home_
-   mailbox: **(I don't know enough about this integration)**
-   media_player: **not off or idle**
_This would then be where a media player is being used in some form and therefore is more notable_
-   notify: **(I don't know what this would be. Notification groups are not in states?)**
-   persistent_notification: **on**
_I don't think there will be much discussion for this one._
-   plant: **problem**
_I don't think there will be much discussion for this one._
-   proximity: **(Should not change)**
-   remote: **(Should not change)**
-   scene: **on**
_I don't think there will be much discussion for this one._
-   script: **running**
_I don't think there will be much discussion for this one._
-   sensor: **(Should not change)**
-   simple_alarm: **I'm not familiar with this one**
-   sun: **above_horizon**
_I don't think there will be much discussion for this one._
-   switch: **on**
_I don't think there will be much discussion for this one._
-   timer: **active/paused**
_Both much different than idle and likely to be addressed_
-   updater: **(I'm not familiar with this one)**
-   vacuum: **cleaning**
-   water_heater: **(Should not change)**
_I'm not sure about this one, but appears as just the operation mode is reported as state_
-   weather: **(Should not change)**
-   weblink: **(Should not change)**
-   zone: **(Should not change)**
-   zwave: **dead**
_I don't know enough about zwave to answer this one for sure_

balloob commented:
I guess I am not convinced that "needing attention" is a good strategy. I still prefer using energy. 

Not all domains will have states. Some are only because they are shown in the logbook. Like Alexa/Google.

Scenes don't change state and can't be colored. Same with zones.

Climate. This one if anything should be linked to the action, not the mode ? And if it's set to mode, all modes should be colored except if it's off. Should follow same coloring as the thermostat card. Consistency is important. We can't have two colors to indicate that a device is heating. 

I don't believe that a person or device tracker being home should be colored.

iantrich commented:
I agree that using energy is a good piece of criteria, but is also too narrow for all of home automation in my opinion.

Climate: I would be on board with aligning with thermo card for hvac_action

SeanPM5 commented:
I lean towards preferring the "requires attention" idea and think ian's list above is mostly good.

Perhaps another way of thinking about this: if you're about to go to sleep/leave home/go on vacation and take a quick look at Home Assistant before doing so, what information is most important to know at a glance? Whether there's any doors left opened or unlocked, whether your alarm is armed, if you forgot to turn off some device, etc.

I don't think device tracker / person should be colored as that one feels too subjective. You could argue that "home" state is more significant as automations depend on that, but many parents would probably think their child being "away" requires more attention.

balloob commented:
I think that if we want to build an interface that is glance worthy in case you're going to sleep/leave home etc, we should probably not conflate that with our control UI?

(we have already discussed a couple of times to get sentence summaries of rooms/everything and is still on some list)

balloob commented:
![image](https://user-images.githubusercontent.com/1444314/72704415-a067eb00-3b0d-11ea-8aa5-da8a603ba1a7.png)

SeanPM5 commented:
I agree that if most icons become yellow it would kinda defeat the point and just tip things too far in the opposite direction. 

But at the same time, the current icon coloring strategy feels somewhat arbitrary? For example:

![icons](https://i.imgur.com/meht2ok.png)

These all have the same exact "on" state, they're all in the same card, they all have toggles, yet some get coloring and others don't. They all look the same, yet behave differently.  Isn't this kinda confusing for new users? I don't know how to easily explain this beyond "that's just the way things are."


One idea I had before was to drop coloring on entity rows altogether. The reasoning being that in most cases there's a colorful blue toggle to quickly determine state, and coloring the icon feels like duplicated functionality. iOS and Android don't color icons when there's a toggle switch. Plus for non-controllable entities, the state is always shown right next to it as text. But I suspect this would be kinda controversial so I was too afraid to ever suggest it publicly 😄

IMO though, this has always been less about the Entities card and more about other cards like Glance and Button. The glance card can be configured with `show_state: false` and really needs coloring in that case. Here is the same entities from above but in glance card form:

![glance](https://i.imgur.com/DjJJneT.png)

Everything is "on" here just like above, yet if you showed anyone this picture, they would say that two things are on and three are off. 

This is even more of an issue on the button card because that doesn't even have a `show_state` option at all. If you put `input_boolean.guest_mode` inside of a button card you'd have no way of knowing whether it was on or off. This makes the button card useless in many cases.

iantrich commented:
The glance and button cards really do highlight the need. I agree with entity rows, to an extent. I don't think we should remove it, but perhaps have an option for the card `color_state_changes` or something.

FWIW, the remote is a bit useless; you can't even turn a remote off to my knowledge

bramkragten commented:
We should at the very least align the different domains with coloring, that binary sensors do color and other domains don't doesn't make sense. We should make sure that a lock/cover binary sensor is the same as the lock and cover domain.

I do agree with Sean that it is mostly the glance/picture elements/button cards that need this, the entities row is not really needed, but I think it is weird to not do it in the entity row for consistency? As you would still want to do it for light to show what color it is?

I gave it some thought and requiring attention might be a slippery slope for a ruleset. It might be different from entity to entity and user to user. Anything that has a clear on/off state could be colored like a switch and binary sensor already are.

But I don't think much of Ian's table will change if we would base it on on/off.

For domains that do not have a clear on/off like climate, for example, we might be able to use other colors to indicate the state. For climate, we use color in the thermostat card to indicate the state, so we could expand those to the rest of the cards.

Eventually, I think we will get settings for this, also for sensor values to be in specific ranges to give it a specific color, etc. But we should have a sane default that works for 90% of the users and use cases.

iantrich commented:
I don't we should touch sensor 😄 
I would much prefer we do something like this for something that has so much variation
```yaml
type: glance
entities:
  - entity: sensor.download_speed
    icon_color: red
    states:
      - value: 10
        operator: '>'
        icon_color: yellow
      - value: 20
        operator: '>'
        icon_color: green
```
How we accomplish that is another discussion though 😛 @balloob and I had previously discussed a state-merge-config card that would allow you to merge a configuration based on the current state of an entity. I think it would be an AWESOME card, but wrapper cards can be quite advanced and there's little chance of a nice UI editor for something like that. Anyways, that leads back to a user being able to define the icon color via the configuration which already I've had push back on.

I think ideally each card just have a state merge function that allowed for such changes, but again, I'm side-tracking myself 😄

bramkragten commented:
Yeah I said eventually, like not in this PR ;-)

balloob commented:
Alright, what about:

 - We implement Ian's proposal more or less.
 - `<state-badge>` gets a new boolean property if it should be `state-colored`. (This is the component that renders the colored icons today.)
 - Entity rows set this default to false.
 - Glance/picture elements set this to True.
 - _Maybe but rather not_: add an option to entities card to enable `state-colored`

Mariusthvdb commented:
> Alright, what about:

> * _Maybe but rather not_: add an option to entities card to enable `state-colored`

First of all let me thank you for discussing this, as a fervent customizer of state colors, I truly welcome the extra options for Lovelace this would offer. So do continue!

Please allow my 2cts on this: I feel using/adding `state-colered:` would be a great addition indeed, and I truly hope you will add it. 

Moreover, might I suggest 1 extra option expanding on that:
- if `state-colered: true`, also add the option for setting the color using `state-color-on:` and `state-color-off:` 
That would allow the user to set the accent-  and off-color easily. (of course not setting it would follow the current theme coloring scheme by default) 

Doing so would keep the global coloring schemes based on theme, And enable incidental coloring per card individually, using outside-theme coloring, which would be a main driver for using the `state-color:` option in the first place.

Please also remember users might want to change the off color also, not only the on color...
thanks.

SeanPM5 commented:
I think balloob's proposal sounds good.

* Makes glance/button/picture elements cards a lot more useful
* Brings consistent behavior to entity row
* Avoids the potential 'sea of yellow icons' problem
* Is very easy to explain. This card doesn't have coloring, these other ones do.
* Seems relatively easy to change behavior later on if needed?

I suspect some will gripe about the entity row changes in the short term simply because they've been used to things working a certain way. Though I think most people will get over it quickly and realize coloring isn't necessary there. 

But might want to make a compromise and still always color lights though in entity rows? Not sure.

balloob commented:
@Mariusthvdb color customizations can happen with a theme-per-row setting. But let's keep that out of this PR.

iantrich committed:
👌 address review comments

iantrich committed:
state_color option for entities card

iantrich commented on src/panels/lovelace/elements/hui-state-icon-element.ts:
```diff
@@ -71,6 +71,7 @@ export class HuiStateIconElement extends LitElement implements LovelaceElement {
           hasAction(this._config.tap_action) ? "0" : undefined
         )}
         .overrideIcon=${this._config.icon}
+        .stateColor="true"
```
add option?

balloob replied:
```suggestion
        statecolor
```

iantrich commented on src/panels/lovelace/cards/hui-glance-card.ts:
```diff
@@ -235,6 +235,7 @@ export class HuiGlanceCard extends LitElement implements LovelaceCard {
                 .stateObj=${stateObj}
                 .overrideIcon=${entityConf.icon}
                 .overrideImage=${entityConf.image}
+                .stateColor="true"
```
add option?

balloob replied:
Let's not start with an option. Like, who wants a glance card without being able to see what's going on. 

However, this is not how you define boolean properties.
```suggestion
                statecolor
```

iantrich committed:
fix typo

balloob commented on src/components/entity/state-badge.ts:
```diff
@@ -16,12 +16,14 @@ import { HassEntity } from "home-assistant-js-websocket";
   @property() public stateObj?: HassEntity;
   @property() public overrideIcon?: string;
   @property() public overrideImage?: string;
+  @property() public stateColor?: boolean;
```
Should it be stateColor or colorState? 

```suggestion
  @property({type: Boolean}) public stateColor?: boolean;
```

iantrich replied:
Is this a recent change to the syntax?

bramkragten replied:
Nope

balloob replied:
Lit runs in the browser when TypeScript has been transpiled away. So `@property` doesn't know the type. By specifying the type for Lit, it will help with serializing/deserializing the value to/from attributes. In the case of boolean it will become "set to true if attribute exists". 

https://github.com/Polymer/lit-element/blob/master/src/lib/updating-element.ts#L138-L164

iantrich replied:
Got it. Thanks for the insight 👍

balloob commented on src/components/entity/state-badge.ts:
```diff
@@ -34,8 +36,8 @@ class StateBadge extends LitElement {
-        data-domain=${computeStateDomain(stateObj)}
-        data-state=${stateObj.state}
+        data-domain=${this.stateColor ? computeStateDomain(stateObj) : ""}
+        data-state=${this.stateColor ? computeActiveState(stateObj) : ""}
```
This breaks unavailable color

balloob commented on src/components/entity/state-badge.ts:
```diff
@@ -34,8 +36,8 @@ class StateBadge extends LitElement {
         id="icon"
-        data-domain=${computeStateDomain(stateObj)}
-        data-state=${stateObj.state}
+        data-domain=${this.stateColor ? computeStateDomain(stateObj) : ""}
```
Use `ifDefined`

balloob commented on src/components/entity/state-badge.ts:
```diff
@@ -111,15 +113,38 @@ class StateBadge extends LitElement {
+      ha-icon[data-domain="timer"][data-state="active"],
+      ha-icon[data-domain="timer"][data-state="paused"],
+      ha-icon[data-domain="vacuum"][data-state="cleaning"],
+      ha-icon[data-domain="zwave"][data-state="dead"] {
```
let's drop this.

balloob commented on src/components/entity/state-badge.ts:
```diff
@@ -111,15 +113,38 @@ class StateBadge extends LitElement {
+      ha-icon[data-domain="lock"][data-state="unlocked"],
+      ha-icon[data-domain="media_player"][data-state="paused"],
+      ha-icon[data-domain="media_player"][data-state="playing"],
+      ha-icon[data-domain="plant"][data-state="problem"],
```
I find it not ideal that "needs attention" is the strategy. I think maybe "doing its thing" is a bette rone. A camera is streaming, the sun is above the horizon, a timer is active. Now there is all of a sudden a problem state in it. Same with Z-Wave. I think that we should not highlight problems, because then people for sure will request that we color them differently !

iantrich replied:
it we add per entity theming, as you suggested, it would be a moot-point on their part.

balloob replied:
Right. But our default UI should still make sense ! If we render errors in red, a dead or problem state should not be yellow ?

iantrich replied:
I've extracted these
```
  ha-icon[data-domain="plant"][data-state="problem"],
  ha-icon[data-domain="zwave"][data-state="dead"] {
    color: var(--error-state-color, #db4437);
  }
```
Is this acceptable?

balloob commented on src/panels/lovelace/cards/hui-entity-button-card.ts:
```diff
@@ -210,14 +211,38 @@ class HuiEntityButtonCard extends LitElement implements LovelaceCard {
 
-      ha-icon[data-domain="light"][data-state="on"],
-      ha-icon[data-domain="switch"][data-state="on"],
+      ha-icon[data-domain="alarm_control_panel"][data-state="disarmed"],
```
Rip out all this CSS from both places and please extract it into a constant.

iantrich replied:
so a single `active/heat/cool/unavailable` attribute?

iantrich committed:
👌 address comments

extract common css
properly set boolean attributes
separate error states/color
fix unavailable color

balloob commented on src/common/style/icon_color_css.ts:
```diff
@@ -0,0 +1,43 @@
+  ha-icon[data-domain="sun"][data-state="above_horizon"],
+  ha-icon[data-domain="switch"][data-state="on"],
+  ha-icon[data-domain="timer"][data-state="active"],
+  ha-icon[data-domain="timer"][data-state="paused"],
```
I would not consider paused to light up.

iantrich replied:
I thought about putting this and media_player paused into their own have having them pulse, but didn't think many would like that

iantrich replied:
![Peek 2020-01-22 12-50](https://user-images.githubusercontent.com/1287159/72924139-0757f080-3d16-11ea-887b-e9ee53da6cb3.gif)

balloob replied:
Nah, that's a lot of attention grabbing. I suggest we keep it uncolored for now.

balloob commented on src/panels/lovelace/cards/hui-entities-card.ts:
```diff
@@ -197,7 +197,10 @@ class HuiEntitiesCard extends LitElement implements LovelaceCard {
 
   private renderEntity(entityConf: EntitiesCardEntityConfig): TemplateResult {
-    const element = createRowElement(entityConf);
+    const element = createRowElement({
```
Let's only make a copy of `entityConf` if `this._config.state_color` is `true` to avoid creating more objects than we need.

iantrich committed:
only make copy if necessary

balloob commented on src/panels/lovelace/cards/hui-entities-card.ts:
```diff
@@ -197,10 +197,14 @@ class HuiEntitiesCard extends LitElement implements LovelaceCard {
+    const element = createRowElement(
+      this._config!.state_color
+        ? {
+            state_color: this._config!.state_color,
```
```suggestion
            state_color: true,
```

balloob commented:
I think that it looks fine from my end. @bramkragten and @SeanPM5, any opinions on coloring?

iantrich committed:
remove paused timer

iantrich committed:
remove paused timer

SeanPM5 commented:
Coloring choices look good to me 👍

iantrich commented:
FYI, colored lights still change in the entities card

balloob commented:
Makes sense colored bulbs are still colored. Should we make non-colored lights turn yellow? 😛

iantrich commented:
> Makes sense colored bulbs are still colored. Should we make non-colored lights turn yellow?

the switch should suffice and a user can override if they wish with `state_color` option
